pfe-badge

diwanshi opened the pull request:
## Title of this pull request

* Link to discovery documentation (should contain all the planning and requirements for this work):
   - See the `DISCOVERY.md` template at the root of the project
   - [https://docs.google.com/document/d/16oPAc5Jfgel9XbwXSWKF28YNdmwpl3KUxQDAnndl7rA/edit#](url)
   - Currently this badge only supports numerical values. For text badges there is a discussion going on around the naming for text badges or making text badges part of pfe-badge component.
   - References to the above discussion:
      - [https://github.com/patternfly/patternfly-unified-design-kit/issues/15](url)
      - [https://github.com/patternfly/patternfly-next/issues/2516](url)
* Link(s) to demo pages where this element can be viewed:

---

### What has changed and why
Summarize files edited as part of this MR along with a brief description of what was changed/why.

* pfe-badge is a new component created to display small numerical UI descriptors.

### Testing instructions
Be sure to include detailed instructions on how your update can be tested by another developer.

1. 

#### Browser requirements
Your component should work in all of the following environments:

- [ ] Latest 2 versions of Edge
- [x] Internet Explorer 11 (should be useable, not pixel perfect)
- [x] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [x] Firefox 60.7.2 (or latest version for Red Hat Enterprise Linux distribution)
- [x] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [x] Latest 2 versions of Safari
- [ ] Galaxy S9 Firefox
- [ ] iPhone X Safari
- [ ] iPad Pro Safari
- [ ] Pixel 3 Chrome

Your repository infrastructure updates should work for at least:
- [ ] Node v8.x
- [ ] NPM v7.x

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [x] Did browser testing pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [x] Was this feature demo'd and the design review approved?
- [ ] Did you update the CHANGELOG.md file with a summary of this update?


**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

diwanshi commented:
@eyevana hey I've forced pushed this branch and made the changes. 
After rebasing, I had to do some changes to the directory structure as the commits had old directory structure.

Krystal has already added some styling so I kept all of those.

eyevana commented:
@diwanshi I applied PF4 styles to the badge. I created local variables, but feel free to replace them our own sass variables.

diwanshi commented:
@eyevana @kylebuch8 hi 
I've updated the pfe-badge with the changes discussed with kyle. I've updated the storybook too.

summary of updates:
* updated the `threshold` attr to `pfe-threshold`
* replaced `read` and `unread` attr with `pfe-state `attr that will have the following states:
  * moderate
  * info
  * critical
  * important
  * success
* added `number` and `text` as new attr that will handle numerical and string values respectively
* updated the styles according to the new syntax for broadcasting and local variables

kylebuch8 commented on elements/pfe-badge/test/pfe-badge_test.html:
```diff
@@ -0,0 +1,27 @@
+    <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes">
+    <script src="/components/@webcomponents/webcomponentsjs/webcomponents-bundle.js"></script>
+    <script src="/components/web-component-tester/browser.js"></script>
+    <script type="module" src="../pfe-badge.js"></script>
```
@diwanshi - Can you update this to `../dist/pfe-badge.js` so the tests start passing?

diwanshi replied:
@kylebuch8 updated.

kylebuch8 commented on elements/pfe-badge/src/pfe-badge.html:
```diff
@@ -0,0 +1,3 @@
+<span class="pfe-badge__container">
+    <slot></slot>
```
I think we should consider getting rid of the slot and make this component more like pfe-number.
https://github.com/patternfly/patternfly-elements/blob/master/elements/pfe-number/src/pfe-number.html

I think the approach that we need to take is to take the value from the `number` or `text` attribute, put it as the `textContent` of the span in `pfe-badge.html` and then update the lightDOM with `this.textContent`.

diwanshi replied:
@kylebuch8 hey I updated the code and now using number or text attribute value to set textcontent of the span

kylebuch8 commented on elements/pfe-badge/src/pfe-badge.js:
```diff
@@ -0,0 +1,54 @@
+      default:
+        return;
+    }
+  }
```
Maybe instead of using a slot in the HTML, we update the textContent of the span in the template and then do something like `this.shadowRoot.querySelector("span").textContent = this.textContent`.

diwanshi replied:
Made this change

kylebuch8 commented on elements/pfe-badge/src/pfe-badge.js:
```diff
@@ -0,0 +1,55 @@
+    super(PfeBadge);   
+  }
+
+  connectedCallback() {
```
Can we delete this line since we're not using it?

kylebuch8 commented on elements/pfe-badge/src/pfe-badge.js:
```diff
@@ -0,0 +1,55 @@
+      default:
+        return;
+    }
+    this.shadowRoot.querySelector("span").textContent = this.textContent;
```
Can you move the `querySelector` to the `constructor` and store it as a variable like `this._textContainer = this.shadowRoot.querySelector("span")` so we're not calling a query selector every time an attribute changes?

diwanshi commented:
@kylebuch8
Added tests and made the span class changes that you mentioned yesterday.

@eyevana i had some time so I thought of adding test cases too. Let me know if they are sufficient.

eyevana commented:
@diwanshi Minor updates from my end. Will continue with remaining documentation items tomorrow. pfe-badge looks freaking amazing! Kudos!! :)

eyevana commented:
@diwanshi after talking with Kyle this morning, it looks like we're going to be building two new components (woohoo! 🎉), pfe-badge for numerical content and [pfe-label](https://www.patternfly.org/v4/documentation/react/components/label) for textual content. I think we're also migrating the various states to pfe-label and leaving pfe-badge with two states (sort of like read and unread). @kylebuch8 can you clarify requirements when you have a chance, pretty please? Thanks, team!

markcaron commented:
NO `pfe-label`! We need to 100% name it something else. My vote is for `pfe-flag`, but if someone has something better, I'm [lit-trally](https://www.youtube.com/watch?v=UgnxlpSQgqg) all ears.

For reference, there was a discussion on this in our [PF Unified Design Kit](https://github.com/patternfly/patternfly-unified-design-kit/issues/15).

eyevana commented:
@markcaron lol, roger that! Let me catch up on the discussion. Thanks for sharing.

kylebuch8 commented on elements/pfe-badge/src/pfe-badge.scss:
```diff
@@ -0,0 +1,50 @@
+  :host([pfe-state="#{$state}"]) {
+    span {
+      --pfe-badge--BackgroundColor: #{pfe-color(feedback--#{$state})};
+      --pfe-badge--Color: #{pfe-color(text--on-dark)};
```
@diwanshi Can we add an if statement here where if the state is "moderate", `--pfe-badge--Color` should be `#{pfe-color(text)}`?

diwanshi commented:
@kylebuch8 i've added the check for moderate state. 

@eyevana i've removed the references to text attribute from readme and storybook

diwanshi commented:
@starryeyez024 hey would you able to review this one as @kylebuch8 is in PTO?
And i think this is needed for Filter list component

kylebuch8 commented:
@diwanshi - I'm taking a look at this right now and will merge it this morning.
